From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/fwctl/mlx5/main.c

---
 drivers/fwctl/mlx5/main.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- a/drivers/fwctl/mlx5/main.c
+++ b/drivers/fwctl/mlx5/main.c
@@ -382,6 +382,7 @@ static int mlx5ctl_probe(struct auxiliar
 	if (ret)
 		return ret;
 	auxiliary_set_drvdata(adev, no_free_ptr(mcdev));
+
 	return 0;
 }
 
@@ -397,7 +398,12 @@ static const struct auxiliary_device_id
 	{.name = MLX5_ADEV_NAME ".fwctl",},
 	{}
 };
+
+#ifdef HAVE_NATIVE_AUX
 MODULE_DEVICE_TABLE(auxiliary, mlx5ctl_id_table);
+#else
+MODULE_ALIAS("auxiliary:" MLX5_ADEV_NAME ".fwctl");
+#endif
 
 static struct auxiliary_driver mlx5ctl_driver = {
 	.name = "mlx5_fwctl",
