From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/fwctl/mlx5/main.c

---
 drivers/fwctl/mlx5/main.c | 6 ++++++
 1 file changed, 6 insertions(+)

--- a/drivers/fwctl/mlx5/main.c
+++ b/drivers/fwctl/mlx5/main.c
@@ -375,7 +375,13 @@ static int mlx5ctl_probe(struct auxiliar
 	ret = fwctl_register(&mcdev->fwctl);
 	if (ret)
 		return ret;
+#ifdef HAVE_AUTO_TYPE
 	auxiliary_set_drvdata(adev, no_free_ptr(mcdev));
+#else
+	auxiliary_set_drvdata(adev, mcdev);
+	mcdev = NULL;
+#endif
+
 	return 0;
 }
 
