From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/addr.c

---
 drivers/infiniband/core/addr.c | 25 ++++++++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

--- a/drivers/infiniband/core/addr.c
+++ b/drivers/infiniband/core/addr.c
@@ -49,6 +49,10 @@
 #include <rdma/ib.h>
 #include <rdma/rdma_netlink.h>
 #include <net/netlink.h>
+#ifdef HAVE_BASECODE_EXTRAS
+#include <net/addrconf.h>
+#include <linux/ratelimit.h>
+#endif
 
 #include "core_priv.h"
 
@@ -165,7 +169,11 @@ static int ib_nl_ip_send_msg(struct rdma
 	}
 
 	/* Construct the family header first */
-	header = skb_put(skb, NLMSG_ALIGN(sizeof(*header)));
+	header =
+#ifdef HAVE_BASECODE_EXTRAS
+		(struct rdma_ls_ip_resolve_header *)
+#endif
+		skb_put(skb, NLMSG_ALIGN(sizeof(*header)));
 	header->ifindex = dev_addr->bound_dev_if;
 	nla_put(skb, attrtype, size, daddr);
 
@@ -237,11 +245,20 @@ rdma_find_ndev_for_src_ip_rcu(struct net
 
 	switch (src_in->sa_family) {
 	case AF_INET:
+#ifdef HAVE___IP_DEV_FIND
 		dev = __ip_dev_find(net,
 				    ((const struct sockaddr_in *)src_in)->sin_addr.s_addr,
 				    false);
 		if (dev)
 			ret = 0;
+#else
+		dev = ip_dev_find(net,
+				   ((const struct sockaddr_in *)src_in)->sin_addr.s_addr);
+		if(dev) {
+			dev_put(dev);
+			ret = 0;
+		}
+#endif
 		break;
 #if IS_ENABLED(CONFIG_IPV6)
 	case AF_INET6:
@@ -406,6 +423,8 @@ static int addr6_resolve(struct sockaddr
 	struct flowi6 fl6;
 	struct dst_entry *dst;
 
+	if (!ipv6_mod_enabled())
+		return -EADDRNOTAVAIL;
 	memset(&fl6, 0, sizeof fl6);
 	fl6.daddr = dst_in->sin6_addr;
 	fl6.saddr = src_in->sin6_addr;
@@ -423,7 +442,7 @@ static int addr6_resolve(struct sockaddr
 	*pdst = dst;
 	return 0;
 }
-#else
+#else /* IS_ENABLED(CONFIG_IPV6) */
 static int addr6_resolve(struct sockaddr *src_sock,
 			 const struct sockaddr *dst_sock,
 			 struct rdma_dev_addr *addr,
@@ -431,7 +450,7 @@ static int addr6_resolve(struct sockaddr
 {
 	return -EADDRNOTAVAIL;
 }
-#endif
+#endif /*IS_ENABLED(CONFIG_IPV6) */
 
 static bool is_dst_local(const struct dst_entry *dst)
 {
