From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/cq.c

---
 drivers/infiniband/core/cq.c | 3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/infiniband/core/cq.c
+++ b/drivers/infiniband/core/cq.c
@@ -8,6 +8,9 @@
 
 #include "core_priv.h"
 
+#ifdef HAVE_BASECODE_EXTRAS
+#include <linux/dim.h>
+#endif
 #include <trace/events/rdma_core.h>
 /* Max size for shared CQ, may require tuning */
 #define IB_MAX_SHARED_CQ_SZ		4096U
