From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/iwcm.c

---
 drivers/infiniband/core/iwcm.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/infiniband/core/iwcm.c b/drivers/infiniband/core/iwcm.c
index xxxxxxx..xxxxxxx 100644
--- a/drivers/infiniband/core/iwcm.c
+++ b/drivers/infiniband/core/iwcm.c
@@ -113,6 +113,9 @@ static struct ctl_table iwcm_ctl_table[] = {
 		.extra1		= SYSCTL_ZERO,
 		.extra2		= SYSCTL_INT_MAX,
 	},
+#ifndef HAVE_REMOVE_SENTINEL_FROM_CTL_TABLE
+	{ }
+#endif
 };
 
 /*
@@ -483,6 +486,7 @@ static void iw_cm_check_wildcard(struct sockaddr_storage *pm_addr,
 			cm4_outaddr->sin_addr = cm4_addr->sin_addr;
 		}
 	} else {
+#if IS_ENABLED(CONFIG_IPV6)
 		struct sockaddr_in6 *pm6_addr = (struct sockaddr_in6 *)pm_addr;
 
 		if (ipv6_addr_type(&pm6_addr->sin6_addr) == IPV6_ADDR_ANY) {
@@ -493,6 +497,7 @@ static void iw_cm_check_wildcard(struct sockaddr_storage *pm_addr,
 
 			cm6_outaddr->sin6_addr = cm6_addr->sin6_addr;
 		}
+#endif
 	}
 }
 
