From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/lag.c

---
 drivers/infiniband/core/lag.c | 4 ++++
 1 file changed, 4 insertions(+)

--- a/drivers/infiniband/core/lag.c
+++ b/drivers/infiniband/core/lag.c
@@ -90,9 +90,13 @@ static struct net_device *rdma_get_xmit_
 		return ERR_PTR(-ENOMEM);
 
 	rcu_read_lock();
+#ifdef HAVE_NETDEV_GET_XMIT_SLAVE
 	slave = netdev_get_xmit_slave(master, skb,
 				      !!(device->lag_flags &
 					 RDMA_LAG_FLAGS_HASH_ALL_SLAVES));
+#else
+	slave = NULL;
+#endif
 	dev_hold(slave);
 	rcu_read_unlock();
 	kfree_skb(skb);
