From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/rdma_core.c

---
 drivers/infiniband/core/rdma_core.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

--- a/drivers/infiniband/core/rdma_core.c
+++ b/drivers/infiniband/core/rdma_core.c
@@ -525,7 +525,6 @@ static void alloc_abort_idr_uobject(stru
 {
 	ib_rdmacg_uncharge(&uobj->cg_obj, uobj->context->device,
 			   RDMACG_RESOURCE_HCA_OBJECT);
-
 	xa_erase(&uobj->ufile->idr, uobj->id);
 }
 
@@ -546,13 +545,13 @@ static int __must_check destroy_hw_idr_u
 
 	ib_rdmacg_uncharge(&uobj->cg_obj, uobj->context->device,
 			   RDMACG_RESOURCE_HCA_OBJECT);
-
 	return 0;
 }
 
 static void remove_handle_idr_uobject(struct ib_uobject *uobj)
 {
 	xa_erase(&uobj->ufile->idr, uobj->id);
+
 	/* Matches the kref in alloc_commit_idr_uobject */
 	uverbs_uobject_put(uobj);
 }
