From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/user_mad.c

---
 drivers/infiniband/core/user_mad.c | 33 ++++++++++++++++++++++++------
 1 file changed, 27 insertions(+), 6 deletions(-)

diff --git a/drivers/infiniband/core/user_mad.c b/drivers/infiniband/core/user_mad.c
index xxxxxxx..xxxxxxx 100644
--- a/drivers/infiniband/core/user_mad.c
+++ b/drivers/infiniband/core/user_mad.c
@@ -354,9 +354,9 @@ static ssize_t copy_recv_mad(struct ib_umad_file *file, char __user *buf,
 				return -EFAULT;
 		}
 	}
-
+#ifndef MLX_DISABLE_TRACEPOINTS
 	trace_ib_umad_read_recv(file, &packet->mad.hdr, &recv_buf->mad->mad_hdr);
-
+#endif
 	return hdr_size(file) + packet->length;
 }
 
@@ -375,10 +375,10 @@ static ssize_t copy_send_mad(struct ib_umad_file *file, char __user *buf,
 
 	if (copy_to_user(buf, packet->mad.data, packet->length))
 		return -EFAULT;
-
+#ifndef MLX_DISABLE_TRACEPOINTS
 	trace_ib_umad_read_send(file, &packet->mad.hdr,
 				(struct ib_mad_hdr *)&packet->mad.data);
-
+#endif
 	return size;
 }
 
@@ -547,8 +547,10 @@ static ssize_t ib_umad_write(struct file *filp, const char __user *buf,
 
 	mutex_lock(&file->mutex);
 
+#ifndef MLX_DISABLE_TRACEPOINTS
 	trace_ib_umad_write(file, &packet->mad.hdr,
 			    (struct ib_mad_hdr *)&packet->mad.data);
+#endif
 
 	agent = __get_agent(file, packet->mad.hdr.id);
 	if (!agent) {
@@ -592,7 +594,7 @@ static ssize_t ib_umad_write(struct file *filp, const char __user *buf,
 	}
 
 	base_version = ((struct ib_mad_hdr *)&packet->mad.data)->base_version;
-	if (check_sub_overflow(count, hdr_size(file) + hdr_len, &data_len)) {
+	if (check_sub_overflow(count, (size_t)(hdr_size(file) + hdr_len), &data_len)) {
 		ret = -EINVAL;
 		goto err_ah;
 	}
@@ -1037,8 +1039,11 @@ static int ib_umad_open(struct inode *inode, struct file *filp)
 	filp->private_data = file;
 
 	list_add_tail(&file->port_list, &port->file_list);
-
+#ifdef HAVE_STREAM_OPEN
 	stream_open(inode, filp);
+#else
+	nonseekable_open(inode, filp);
+#endif
 out:
 	mutex_unlock(&port->file_mutex);
 	return ret;
@@ -1089,6 +1094,9 @@ static const struct file_operations umad_fops = {
 #endif
 	.open		= ib_umad_open,
 	.release	= ib_umad_close,
+#ifdef HAVE_NO_LLSEEK
+	.llseek		= no_llseek,
+#endif
 };
 
 static int ib_umad_sm_open(struct inode *inode, struct file *filp)
@@ -1156,6 +1164,9 @@ static const struct file_operations umad_sm_fops = {
 	.owner	 = THIS_MODULE,
 	.open	 = ib_umad_sm_open,
 	.release = ib_umad_sm_close,
+#ifdef HAVE_NO_LLSEEK
+	.llseek	 = no_llseek,
+#endif
 };
 
 static struct ib_umad_port *get_port(struct ib_device *ibdev,
@@ -1243,13 +1254,22 @@ static struct attribute *umad_class_dev_attrs[] = {
 };
 ATTRIBUTE_GROUPS(umad_class_dev);
 
+#ifdef HAVE_DEVNODE_GET_CONST_DEVICE
 static char *umad_devnode(const struct device *dev, umode_t *mode)
+#else
+static char *umad_devnode(struct device *dev, umode_t *mode)
+#endif
 {
 	return kasprintf(GFP_KERNEL, "infiniband/%s", dev_name(dev));
 }
 
+#ifdef HAVE_SHOW_CLASS_ATTR_STRING_GET_CONST
 static ssize_t abi_version_show(const struct class *class,
 				const struct class_attribute *attr, char *buf)
+#else
+static ssize_t abi_version_show(struct class *class,
+				struct class_attribute *attr, char *buf)
+#endif /* HAVE_SHOW_CLASS_ATTR_STRING_GET_CONST */
 {
 	return sysfs_emit(buf, "%d\n", IB_USER_MAD_ABI_VERSION);
 }
@@ -1268,6 +1288,7 @@ static struct class umad_class = {
 	.dev_groups	= umad_class_dev_groups,
 };
 
+
 static void ib_umad_release_port(struct device *device)
 {
 	struct ib_umad_port *port = dev_get_drvdata(device);
