From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/uverbs_ioctl.c

---
 drivers/infiniband/core/uverbs_ioctl.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/infiniband/core/uverbs_ioctl.c b/drivers/infiniband/core/uverbs_ioctl.c
index xxxxxxx..xxxxxxx 100644
--- a/drivers/infiniband/core/uverbs_ioctl.c
+++ b/drivers/infiniband/core/uverbs_ioctl.c
@@ -569,7 +569,6 @@ static int ib_uverbs_cmd_verbs(struct ib_uverbs_file *ufile,
 
 	if (unlikely(hdr->driver_id != uapi->driver_id))
 		return -EINVAL;
-
 	slot = radix_tree_iter_lookup(
 		&uapi->radix, &attrs_iter,
 		uapi_key_obj(hdr->object_id) |
