From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/uverbs_main.c

---
 drivers/infiniband/core/uverbs_main.c | 40 +++++++++++++++++++++++----
 1 file changed, 35 insertions(+), 5 deletions(-)

--- a/drivers/infiniband/core/uverbs_main.c
+++ b/drivers/infiniband/core/uverbs_main.c
@@ -29,6 +29,7 @@
  * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
  * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ 
  * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
@@ -79,14 +80,18 @@ static int ib_uverbs_add_one(struct ib_d
 static void ib_uverbs_remove_one(struct ib_device *device, void *client_data);
 static struct ib_client uverbs_client;
 
+#ifdef HAVE_DEVNODE_GET_CONST_DEVICE
 static char *uverbs_devnode(const struct device *dev, umode_t *mode)
+#else
+static char *uverbs_devnode(struct device *dev, umode_t *mode)
+#endif
 {
 	if (mode)
 		*mode = 0666;
 	return kasprintf(GFP_KERNEL, "infiniband/%s", dev_name(dev));
 }
 
-static const struct class uverbs_class = {
+static struct class uverbs_class = {
 	.name = "infiniband_verbs",
 	.devnode = uverbs_devnode,
 };
@@ -103,7 +108,6 @@ struct ib_ucontext *ib_uverbs_get_uconte
 	 * get_context, we get NULL or valid ucontext.
 	 */
 	struct ib_ucontext *ucontext = smp_load_acquire(&ufile->ucontext);
-
 	if (!srcu_dereference(ufile->device->ib_dev,
 			      &ufile->device->disassociate_srcu))
 		return ERR_PTR(-EIO);
@@ -356,14 +360,20 @@ const struct file_operations uverbs_even
 	.poll    = ib_uverbs_comp_event_poll,
 	.release = uverbs_uobject_fd_release,
 	.fasync  = ib_uverbs_comp_event_fasync,
+#ifdef HAVE_NO_LLSEEK
+	.llseek	 = no_llseek,
+#endif
 };
 
 const struct file_operations uverbs_async_event_fops = {
 	.owner	 = THIS_MODULE,
 	.read	 = ib_uverbs_async_event_read,
 	.poll    = ib_uverbs_async_event_poll,
-	.release = uverbs_async_event_release,
 	.fasync  = ib_uverbs_async_event_fasync,
+	.release = uverbs_async_event_release,
+#ifdef HAVE_NO_LLSEEK
+	.llseek	 = no_llseek,
+#endif
 };
 
 void ib_uverbs_comp_handler(struct ib_cq *cq, void *cq_context)
@@ -820,7 +830,6 @@ static vm_fault_t rdma_umap_fault(struct
 
 	return ret;
 }
-
 static const struct vm_operations_struct rdma_umap_ops = {
 	.open = rdma_umap_open,
 	.close = rdma_umap_close,
@@ -868,7 +877,11 @@ void uverbs_user_mmap_disassociate(struc
 		 * at a time to get the lock ordering right. Typically there
 		 * will only be one mm, so no big deal.
 		 */
+#ifdef HAVE_MMAP_READ_LOCK
 		mmap_read_lock(mm);
+#else
+		down_read(&mm->mmap_sem);
+#endif
 		mutex_lock(&ufile->umap_lock);
 		list_for_each_entry_safe (priv, next_priv, &ufile->umaps,
 					  list) {
@@ -887,7 +900,11 @@ void uverbs_user_mmap_disassociate(struc
 			}
 		}
 		mutex_unlock(&ufile->umap_lock);
+#ifdef HAVE_MMAP_READ_LOCK
 		mmap_read_unlock(mm);
+#else
+		up_read(&mm->mmap_sem);
+#endif
 		mmput(mm);
 	}
 
@@ -994,7 +1011,11 @@ static int ib_uverbs_open(struct inode *
 
 	setup_ufile_idr_uobject(file);
 
+#ifdef HAVE_STREAM_OPEN
 	return stream_open(inode, filp);
+#else
+	return nonseekable_open(inode, filp);
+#endif
 
 err_module:
 	module_put(ib_dev->ops.owner);
@@ -1029,6 +1050,9 @@ static const struct file_operations uver
 	.write	 = ib_uverbs_write,
 	.open	 = ib_uverbs_open,
 	.release = ib_uverbs_close,
+#ifdef HAVE_NO_LLSEEK
+	.llseek	 = no_llseek,
+#endif
 	.unlocked_ioctl = ib_uverbs_ioctl,
 	.compat_ioctl = compat_ptr_ioctl,
 };
@@ -1039,6 +1063,9 @@ static const struct file_operations uver
 	.mmap    = ib_uverbs_mmap,
 	.open	 = ib_uverbs_open,
 	.release = ib_uverbs_close,
+#ifdef HAVE_NO_LLSEEK
+	.llseek	 = no_llseek,
+#endif
 	.unlocked_ioctl = ib_uverbs_ioctl,
 	.compat_ioctl = compat_ptr_ioctl,
 };
@@ -1243,7 +1270,6 @@ static void ib_uverbs_free_hw_resources(
 
 		uverbs_destroy_ufile_hw(file, RDMA_REMOVE_DRIVER_REMOVE);
 		kref_put(&file->ref, ib_uverbs_release_file);
-
 		mutex_lock(&uverbs_dev->lists_mutex);
 	}
 	mutex_unlock(&uverbs_dev->lists_mutex);
@@ -1347,7 +1373,11 @@ static void __exit ib_uverbs_cleanup(voi
 	unregister_chrdev_region(dynamic_uverbs_dev,
 				 IB_UVERBS_NUM_DYNAMIC_MINOR);
 	ib_cleanup_ucaps();
+#ifdef CONFIG_INFINIBAND_ON_DEMAND_PAGING
+#ifdef HAVE_MMU_NOTIFIER_OPS_HAS_FREE_NOTIFIER
 	mmu_notifier_synchronize();
+#endif
+#endif
 }
 
 module_init(ib_uverbs_init);
