From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/hw/mlx5/doorbell.c

---
 drivers/infiniband/hw/mlx5/doorbell.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/infiniband/hw/mlx5/doorbell.c b/drivers/infiniband/hw/mlx5/doorbell.c
index xxxxxxx..xxxxxxx 100644
--- a/drivers/infiniband/hw/mlx5/doorbell.c
+++ b/drivers/infiniband/hw/mlx5/doorbell.c
@@ -124,8 +124,13 @@ static int mlx5_ib_db_map_user_desc(struct mlx5_ib_ucontext *context,
 	list_add(&page->list, &context->db_page_list);
 
 found:
+#ifdef HAVE_SG_APPEND_TABLE
 	db->dma = sg_dma_address(page->umem->sgt_append.sgt.sgl) +
 		  (desc->addr & ~PAGE_MASK);
+#else
+	db->dma = sg_dma_address(page->umem->sg_head.sgl) +
+		  (desc->addr & ~PAGE_MASK);
+#endif
 	db->u.user_page = page;
 	++page->refcnt;
 
