From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/hw/mlx5/umr.c

---
 drivers/infiniband/hw/mlx5/umr.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

--- a/drivers/infiniband/hw/mlx5/umr.c
+++ b/drivers/infiniband/hw/mlx5/umr.c
@@ -695,8 +695,8 @@ _mlx5r_umr_update_mr_pas(struct mlx5_ib_
 	size_t processed_blocks = 0;
 	struct ib_block_iter biter;
 	size_t cur_block_idx = 0;
-	struct mlx5_ksm *cur_ksm;
-	struct mlx5_mtt *cur_mtt;
+	struct mlx5_ksm *cur_ksm = NULL;
+	struct mlx5_mtt *cur_mtt = NULL;
 	size_t orig_sg_length;
 	size_t total_blocks;
 	size_t final_size;
@@ -905,6 +905,7 @@ int mlx5r_umr_update_xlt(struct mlx5_ib_
 		size_to_map = npages * desc_size;
 		dma_sync_single_for_cpu(ddev, sg.addr, sg.length,
 					DMA_TO_DEVICE);
+#ifdef HAVE_HMM_PFN_DMA_MAPPED
 		/*
 		 * npages is the maximum number of pages to map, but we
 		 * can't guarantee that all pages are actually mapped.
@@ -916,6 +917,9 @@ int mlx5r_umr_update_xlt(struct mlx5_ib_
 		err = mlx5_odp_populate_xlt(xlt, idx, npages, mr, flags);
 		if (err)
 			return err;
+#else
+		mlx5_odp_populate_xlt(xlt, idx, npages, mr, flags);
+#endif
 		dma_sync_single_for_device(ddev, sg.addr, sg.length,
 					   DMA_TO_DEVICE);
 		sg.length = ALIGN(size_to_map, MLX5_UMR_FLEX_ALIGNMENT);
