From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/nvme/host/pr.c

---
 drivers/nvme/host/pr.c | 26 ++++++++++++++++++++++++--
 1 file changed, 24 insertions(+), 2 deletions(-)

diff --git a/drivers/nvme/host/pr.c b/drivers/nvme/host/pr.c
index xxxxxxx..xxxxxxx 100644
--- a/drivers/nvme/host/pr.c
+++ b/drivers/nvme/host/pr.c
@@ -29,6 +29,7 @@ static enum nvme_pr_type nvme_pr_type_from_blk(enum pr_type type)
 	return 0;
 }
 
+#ifdef HAVE_PR_KEYS
 static enum pr_type block_pr_type_from_nvme(enum nvme_pr_type type)
 {
 	switch (type) {
@@ -48,6 +49,7 @@ static enum pr_type block_pr_type_from_nvme(enum nvme_pr_type type)
 
 	return 0;
 }
+#endif
 
 static int nvme_send_ns_head_pr_command(struct block_device *bdev,
 		struct nvme_command *c, void *data, unsigned int data_len)
@@ -72,6 +74,7 @@ static int nvme_send_ns_pr_command(struct nvme_ns *ns, struct nvme_command *c,
 	return nvme_submit_sync_cmd(ns->queue, c, data, data_len);
 }
 
+#ifdef HAVE_PR_STATUS
 static int nvme_status_to_pr_err(int status)
 {
 	if (nvme_is_path_error(status))
@@ -91,6 +94,7 @@ static int nvme_status_to_pr_err(int status)
 		return PR_STS_IOERR;
 	}
 }
+#endif
 
 static int __nvme_send_pr_command(struct block_device *bdev, u32 cdw10,
 		u32 cdw11, u8 op, void *data, unsigned int data_len)
@@ -113,7 +117,14 @@ static int nvme_send_pr_command(struct block_device *bdev, u32 cdw10, u32 cdw11,
 	int ret;
 
 	ret = __nvme_send_pr_command(bdev, cdw10, cdw11, op, data, data_len);
-	return ret < 0 ? ret : nvme_status_to_pr_err(ret);
+#ifdef HAVE_PR_STATUS
+	if (ret < 0)
+		return ret;
+
+	return nvme_status_to_pr_err(ret);
+#else
+	return ret;
+#endif
 }
 
 static int nvme_pr_register(struct block_device *bdev, u64 old_key, u64 new_key,
@@ -202,6 +213,7 @@ static int nvme_pr_release(struct block_device *bdev, u64 key, enum pr_type type
 			&data, sizeof(data));
 }
 
+#ifdef HAVE_PR_KEYS
 static int nvme_pr_resv_report(struct block_device *bdev, void *data,
 		u32 data_len, bool *eds)
 {
@@ -222,7 +234,14 @@ retry:
 		goto retry;
 	}
 
-	return ret < 0 ? ret : nvme_status_to_pr_err(ret);
+#ifdef HAVE_PR_STATUS
+	if (ret < 0)
+		return ret;
+
+	return nvme_status_to_pr_err(ret);
+#else
+	return ret;
+#endif
 }
 
 static int nvme_pr_read_keys(struct block_device *bdev,
@@ -328,6 +347,7 @@ free_rse:
 	kfree(rse);
 	return ret;
 }
+#endif
 
 const struct pr_ops nvme_pr_ops = {
 	.pr_register	= nvme_pr_register,
@@ -335,6 +355,8 @@ const struct pr_ops nvme_pr_ops = {
 	.pr_release	= nvme_pr_release,
 	.pr_preempt	= nvme_pr_preempt,
 	.pr_clear	= nvme_pr_clear,
+#ifdef HAVE_PR_KEYS
 	.pr_read_keys	= nvme_pr_read_keys,
 	.pr_read_reservation = nvme_pr_read_reservation,
+#endif
 };
