From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/nvme/target/rdma_offload.c

---
 drivers/nvme/target/rdma_offload.c | 37 ++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

--- a/drivers/nvme/target/rdma_offload.c
+++ b/drivers/nvme/target/rdma_offload.c
@@ -16,6 +16,10 @@
  */
 
 #include "rdma_offload.h"
+#if defined(CONFIG_PPC) && defined(HAVE_PNV_PCI_SET_P2P)
+#include <asm/pnv-pci.h>
+#endif
+#include <linux/sizes.h>
 
 static unsigned int
 __nvmet_rdma_peer_to_peer_sqe_inline_size(struct ib_nvmf_caps *nvmf_caps,
@@ -428,7 +432,17 @@ static void nvmet_rdma_free_be_ctrl(stru
 	bool has_err = false;
 	int ret;
 
+#if defined(CONFIG_PPC) && defined(HAVE_PNV_PCI_SET_P2P)
+	pnv_pci_set_p2p(container_of(xrq->ndev->device->dev.parent,
+				     struct pci_dev, dev),
+			be_ctrl->pdev, OPAL_PCI_P2P_STORE);
+#endif
+#ifdef HAVE_IDA_SIMPLE_REMOVE 
 	ida_simple_remove(&nvmet_rdma_bectrl_ida, be_ctrl->offload_ctx.id);
+#else
+	ida_free(&nvmet_rdma_bectrl_ida, be_ctrl->offload_ctx.id);
+#endif
+
 	nvmet_offload_ctx_configfs_del(&be_ctrl->offload_ctx);
 
 	if (be_ctrl->ibns) {
@@ -629,6 +643,15 @@ nvmet_rdma_create_be_ctrl(struct nvmet_r
 	}
 	mutex_unlock(&xrq->be_mutex);
 
+#if defined(CONFIG_PPC) && defined(HAVE_PNV_PCI_SET_P2P)
+	err = pnv_pci_set_p2p(container_of(xrq->ndev->device->dev.parent,
+					   struct pci_dev, dev),
+			      ns->pdev,
+			      OPAL_PCI_P2P_STORE | OPAL_PCI_P2P_ENABLE);
+	if (err)
+		return ERR_PTR(err);
+#endif
+
 	be_ctrl = kzalloc(sizeof(*be_ctrl), GFP_KERNEL);
 	if (!be_ctrl) {
 		err = -ENOMEM;
@@ -693,8 +716,13 @@ nvmet_rdma_create_be_ctrl(struct nvmet_r
 	be_ctrl->offload_ctx.ctx = be_ctrl;
 	be_ctrl->offload_ctx.port = xrq->port;
 	be_ctrl->offload_ctx.ns = ns;
+#ifdef HAVE_IDA_SIMPLE_REMOVE 
 	be_ctrl->offload_ctx.id = ida_simple_get(&nvmet_rdma_bectrl_ida, 0, 0,
 						 GFP_KERNEL);
+#else
+	be_ctrl->offload_ctx.id = ida_alloc(&nvmet_rdma_bectrl_ida,
+					    GFP_KERNEL);
+#endif
 	if (be_ctrl->offload_ctx.id < 0) {
 		err = -ENOMEM;
 		goto out_detach_ns;
@@ -712,7 +740,11 @@ nvmet_rdma_create_be_ctrl(struct nvmet_r
 	return be_ctrl;
 
 out_ida_remove:
+#ifdef HAVE_IDA_SIMPLE_REMOVE 
 	ida_simple_remove(&nvmet_rdma_bectrl_ida, be_ctrl->offload_ctx.id);
+#else
+	ida_free(&nvmet_rdma_bectrl_ida, be_ctrl->offload_ctx.id);
+#endif
 out_detach_ns:
 	ib_detach_nvmf_ns(be_ctrl->ibns);
 out_destroy_be_ctrl:
@@ -728,6 +760,11 @@ out_free_be_ctrl:
 	kref_put(&xrq->ref, nvmet_rdma_destroy_xrq);
 	kfree(be_ctrl);
 out_err:
+#if defined(CONFIG_PPC) && defined(HAVE_PNV_PCI_SET_P2P)
+	pnv_pci_set_p2p(container_of(xrq->ndev->device->dev.parent,
+				     struct pci_dev, dev),
+			ns->pdev, OPAL_PCI_P2P_STORE);
+#endif
 	return ERR_PTR(err);
 }
 
