From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: include/rdma/rdma_counter.h

---
 include/rdma/rdma_counter.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/include/rdma/rdma_counter.h
+++ b/include/rdma/rdma_counter.h
@@ -6,9 +6,10 @@
 #ifndef _RDMA_COUNTER_H_
 #define _RDMA_COUNTER_H_
 
+#include "../../compat/config.h"
+
 #include <linux/mutex.h>
 #include <linux/pid_namespace.h>
-
 #include <rdma/restrack.h>
 #include <rdma/rdma_netlink.h>
 
