From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: net/mlxdevm/netlink_gen.c

---
 net/mlxdevm/netlink_gen.c | 315 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 315 insertions(+)

--- a/net/mlxdevm/netlink_gen.c
+++ b/net/mlxdevm/netlink_gen.c
@@ -7,6 +7,9 @@
 #include <net/genetlink.h>
 
 #include "netlink_gen.h"
+#ifndef HAVE_GENL_OPS_MAXATTR
+#include "devl_internal.h"
+#endif
 
 #include <uapi/linux/devlink.h>
 
@@ -582,6 +585,7 @@ static const struct nla_policy devlink_n
 #endif
 
 /* Ops table for mlxdevm */
+#ifdef HAVE_STRUCT_GENL_SPLIT_OPS
 const struct genl_split_ops mlxdevm_nl_ops[29] = {
 	{
 		.cmd		= MLXDEVM_CMD_GET,
@@ -1274,3 +1278,314 @@ const struct genl_split_ops mlxdevm_nl_o
 	},
 #endif
 };
+#else
+const struct genl_ops mlxdevm_nl_ops[22] = {
+	{
+		.cmd		= MLXDEVM_CMD_GET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_get_doit,
+		.dumpit		= mlxdevm_nl_get_dumpit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_get_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_DEV_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+	},
+	{
+		.cmd		= MLXDEVM_CMD_PORT_GET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_port_get_doit,
+		.dumpit		= mlxdevm_nl_port_get_dumpit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_port_get_do_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_PORT_INDEX,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+	},
+	{
+		.cmd		= MLXDEVM_CMD_PORT_SET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_port_set_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_port_set_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_PORT_FUNCTION,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_PORT_NEW,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_port_new_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_port_new_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_PORT_PCI_SF_NUMBER,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_PORT_DEL,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_port_del_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_port_del_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_PORT_INDEX,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_ESWITCH_GET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_eswitch_get_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_eswitch_get_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_DEV_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_ESWITCH_SET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_eswitch_set_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_eswitch_set_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_ESWITCH_ENCAP_MODE,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_DPIPE_TABLE_GET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_dpipe_table_get_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_dpipe_table_get_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_DPIPE_TABLE_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+	},
+	{
+		.cmd		= MLXDEVM_CMD_RESOURCE_DUMP,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_resource_dump_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_resource_dump_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_DEV_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+	},
+	{
+		.cmd		= MLXDEVM_CMD_RELOAD,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_reload_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_reload_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_RELOAD_LIMITS,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_PARAM_GET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_param_get_doit,
+		.dumpit		= mlxdevm_nl_param_get_dumpit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_param_get_do_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_PARAM_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+	},
+	{
+		.cmd		= MLXDEVM_CMD_PARAM_SET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_param_set_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_param_set_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_EXT_PARAM_ARRAY_TYPE,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_INFO_GET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_info_get_doit,
+		.dumpit		= mlxdevm_nl_info_get_dumpit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_info_get_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_DEV_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+	},
+	{
+		.cmd		= MLXDEVM_CMD_FLASH_UPDATE,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_flash_update_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_flash_update_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_FLASH_UPDATE_OVERWRITE_MASK,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_TRAP_GET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_trap_get_doit,
+		.dumpit		= mlxdevm_nl_trap_get_dumpit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_trap_get_do_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_TRAP_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+	},
+	{
+		.cmd		= MLXDEVM_CMD_TRAP_SET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_trap_set_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_trap_set_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_TRAP_ACTION,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_TRAP_GROUP_GET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_trap_group_get_doit,
+		.dumpit		= mlxdevm_nl_trap_group_get_dumpit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_trap_group_get_do_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_TRAP_GROUP_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_CMD_CAP_DO,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_TRAP_GROUP_SET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT | GENL_DONT_VALIDATE_DUMP,
+#endif
+		.doit		= mlxdevm_nl_trap_group_set_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_trap_group_set_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_TRAP_POLICER_ID,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_RATE_GET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT,
+#endif
+		.doit		= mlxdevm_nl_rate_get_doit,
+		.dumpit		= mlxdevm_nl_rate_get_dumpit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_rate_get_do_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_RATE_NODE_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+	},
+	{
+		.cmd		= MLXDEVM_CMD_RATE_SET,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT,
+#endif
+		.doit		= mlxdevm_nl_rate_set_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_rate_set_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_RATE_TX_WEIGHT,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_RATE_NEW,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT,
+#endif
+		.doit		= mlxdevm_nl_rate_new_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_rate_new_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_RATE_TX_WEIGHT,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+	{
+		.cmd		= MLXDEVM_CMD_RATE_DEL,
+#ifdef HAVE_GENL_OPS_VALIDATE
+		.validate	= GENL_DONT_VALIDATE_STRICT,
+#endif
+		.doit		= mlxdevm_nl_rate_del_doit,
+#ifdef HAVE_GENL_OPS_MAXATTR
+		.policy		= mlxdevm_rate_del_nl_policy,
+		.maxattr	= MLXDEVM_ATTR_RATE_NODE_NAME,
+#elif defined(HAVE_GENL_OPS_POLICY)
+		.policy		= mlxdevm_nl_policy,
+#endif
+		.flags		= GENL_ADMIN_PERM,
+	},
+};
+#endif
