From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: include/rdma/ib_verbs.h

---
 include/rdma/ib_verbs.h | 30 ++++++++++++++++++++++++++++--
 1 file changed, 28 insertions(+), 2 deletions(-)

diff --git a/include/rdma/ib_verbs.h b/include/rdma/ib_verbs.h
index xxxxxxx..xxxxxxx 100644
--- a/include/rdma/ib_verbs.h
+++ b/include/rdma/ib_verbs.h
@@ -12,6 +12,8 @@
 #ifndef IB_VERBS_H
 #define IB_VERBS_H
 
+#include "../../compat/config.h"
+
 #include <linux/ethtool.h>
 #include <linux/types.h>
 #include <linux/device.h>
@@ -83,10 +85,13 @@ void ibdev_notice(const struct ib_device *ibdev, const char *format, ...);
 __printf(2, 3) __cold
 void ibdev_info(const struct ib_device *ibdev, const char *format, ...);
 
-#if defined(CONFIG_DYNAMIC_DEBUG) || \
+#if defined(CONFIG_DYNAMIC_DEBUG) && defined(dynamic_ibdev_dbg)
+#define ibdev_dbg(__dev, format, args...)                       \
+	        dynamic_ibdev_dbg(__dev, format, ##args)
+#elif defined(DEBUG)
 	(defined(CONFIG_DYNAMIC_DEBUG_CORE) && defined(DYNAMIC_DEBUG_MODULE))
 #define ibdev_dbg(__dev, format, args...)                       \
-	dynamic_ibdev_dbg(__dev, format, ##args)
+	        ibdev_printk(KERN_DEBUG, __dev, format, ##args)
 #else
 __printf(2, 3) __cold
 static inline
@@ -1577,6 +1582,7 @@ struct ib_ucontext {
 	struct ib_device       *device;
 	struct ib_uverbs_file  *ufile;
 
+
 	struct ib_rdmacg_object	cg_obj;
 	u64 enabled_caps;
 	/*
@@ -2286,7 +2292,9 @@ struct ib_port_data {
 	struct ib_port_cache cache;
 
 	struct net_device __rcu *netdev;
+#ifdef HAVE_NETDEV_PUT_AND_HOLD
 	netdevice_tracker netdev_tracker;
+#endif
 	struct hlist_node ndev_hash_link;
 	struct rdma_port_counter port_counter;
 	struct ib_port *sysfs;
@@ -2399,9 +2407,11 @@ struct rdma_user_mmap_entry {
 	unsigned long start_pgoff;
 	size_t npages;
 	bool driver_removed;
+#ifdef HAVE_MMIO_REGION_EXPORTED_THROUGH_DMA_BUF
 	/* protects access to dmabufs */
 	struct mutex dmabufs_lock;
 	struct list_head dmabufs;
+#endif
 };
 
 /* Return the offset (in bytes) the user should pass to libc's mmap() */
@@ -2540,11 +2550,17 @@ struct ib_device_ops {
 	 * Therefore needs to be implemented by the driver in mmap_free.
 	 */
 	void (*mmap_free)(struct rdma_user_mmap_entry *entry);
+#ifdef HAVE_MMIO_REGION_EXPORTED_THROUGH_DMA_BUF
 	int (*mmap_get_pfns)(struct rdma_user_mmap_entry *entry,
+#ifdef HAVE_PHYS_VEC
 			     struct phys_vec *phys_vec,
+#else
+			     struct dma_buf_phys_vec *phys_vec,
+#endif
 			     struct p2pdma_provider **provider);
 	struct rdma_user_mmap_entry *(*pgoff_to_mmap_entry)(struct ib_ucontext *ucontext,
 							    off_t pg_off);
+#endif
 	void (*disassociate_ucontext)(struct ib_ucontext *ibcontext);
 	int (*alloc_pd)(struct ib_pd *pd, struct ib_udata *udata);
 	int (*dealloc_pd)(struct ib_pd *pd, struct ib_udata *udata);
@@ -2626,6 +2642,10 @@ struct ib_device_ops {
 			       struct ib_mr_status *mr_status);
 	int (*alloc_mw)(struct ib_mw *mw, struct ib_udata *udata);
 	int (*dealloc_mw)(struct ib_mw *mw);
+#ifndef HAVE_MMU_INTERVAL_NOTIFIER
+	void (*invalidate_range)(struct ib_umem_odp *umem_odp,
+				 unsigned long start, unsigned long end);
+#endif
 	int (*attach_mcast)(struct ib_qp *qp, union ib_gid *gid, u16 lid);
 	int (*detach_mcast)(struct ib_qp *qp, union ib_gid *gid, u16 lid);
 	int (*alloc_xrcd)(struct ib_xrcd *xrcd, struct ib_udata *udata);
@@ -4224,7 +4244,11 @@ static inline bool ib_dma_pci_p2p_dma_supported(struct ib_device *dev)
 	if (ib_uses_virt_dma(dev))
 		return false;
 
+#ifdef HAVE_DMA_PCI_P2PDMA_SUPPORTED
 	return dma_pci_p2pdma_supported(dev->dma_device);
+#else
+	return false;
+#endif
 }
 
 /**
@@ -4370,6 +4394,7 @@ static inline void ib_dma_unmap_sg_attrs(struct ib_device *dev,
  * @direction: The direction of the DMA
  * @attrs: Optional DMA attributes for the map operation
  */
+#ifdef HAVE_DMA_MAP_SGTABLE
 static inline int ib_dma_map_sgtable_attrs(struct ib_device *dev,
 					   struct sg_table *sgt,
 					   enum dma_data_direction direction,
@@ -4395,6 +4420,7 @@ static inline void ib_dma_unmap_sgtable_attrs(struct ib_device *dev,
 	if (!ib_uses_virt_dma(dev))
 		dma_unmap_sgtable(dev->dma_device, sgt, direction, dma_attrs);
 }
+#endif
 
 /**
  * ib_dma_map_sg - Map a scatter/gather list to DMA addresses
