From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: net/mlxdevm/core.c

---
 net/mlxdevm/core.c | 8 ++++++++
 1 file changed, 8 insertions(+)

--- a/net/mlxdevm/core.c
+++ b/net/mlxdevm/core.c
@@ -347,7 +347,9 @@ static void mlxdevm_release(struct work_
 	mlxdevm = container_of(to_rcu_work(work), struct mlxdevm, rwork);
 
 	mutex_destroy(&mlxdevm->lock);
+#ifdef HAVE_LOCKDEP_UNREGISTER_KEY
 	lockdep_unregister_key(&mlxdevm->lock_key);
+#endif
 	kfree(mlxdevm);
 }
 
@@ -413,7 +415,13 @@ int devm_register(struct mlxdevm *mlxdev
 	INIT_LIST_HEAD(&mlxdevm->trap_group_list);
 	INIT_LIST_HEAD(&mlxdevm->trap_policer_list);
 	INIT_RCU_WORK(&mlxdevm->rwork, mlxdevm_release);
+#ifndef HAVE_DEVLINK_HAS_RATE_FUNCTIONS
+	INIT_LIST_HEAD(&mlxdevm->rate_node_list);
+	init_rwsem(&mlxdevm->rate_node_rwsem);
+#endif
+#ifdef HAVE_LOCKDEP_UNREGISTER_KEY
 	lockdep_register_key(&mlxdevm->lock_key);
+#endif
 	lockdep_set_class(&mlxdevm->lock, &mlxdevm->lock_key);
 	refcount_set(&mlxdevm->refcount, 1);
 
