From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: net/mlxdevm/core.c

---
 net/mlxdevm/core.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/mlxdevm/core.c b/net/mlxdevm/core.c
index xxxxxxx..xxxxxxx 100644
--- a/net/mlxdevm/core.c
+++ b/net/mlxdevm/core.c
@@ -413,6 +413,10 @@ int devm_register(struct mlxdevm *mlxdevm)
 	INIT_LIST_HEAD(&mlxdevm->trap_group_list);
 	INIT_LIST_HEAD(&mlxdevm->trap_policer_list);
 	INIT_RCU_WORK(&mlxdevm->rwork, mlxdevm_release);
+#ifndef HAVE_DEVLINK_HAS_RATE_FUNCTIONS
+	INIT_LIST_HEAD(&mlxdevm->rate_node_list);
+	init_rwsem(&mlxdevm->rate_node_rwsem);
+#endif
 	lockdep_register_key(&mlxdevm->lock_key);
 	lockdep_set_class(&mlxdevm->lock, &mlxdevm->lock_key);
 	refcount_set(&mlxdevm->refcount, 1);
