From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/iwcm.c

---
 drivers/infiniband/core/iwcm.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/drivers/infiniband/core/iwcm.c
+++ b/drivers/infiniband/core/iwcm.c
@@ -58,6 +58,9 @@
 MODULE_AUTHOR("Tom Tucker");
 MODULE_DESCRIPTION("iWARP CM");
 MODULE_LICENSE("Dual BSD/GPL");
+#ifdef RETPOLINE_MLNX
+MODULE_INFO(retpoline, "Y");
+#endif
 
 static const char * const iwcm_rej_reason_strs[] = {
 	[ECONNRESET]			= "reset by remote host",
@@ -110,9 +113,14 @@ static struct ctl_table iwcm_ctl_table[]
 		.maxlen		= sizeof(default_backlog),
 		.mode		= 0644,
 		.proc_handler	= proc_dointvec_minmax,
+#ifdef HAVE_SYSCTL_ZERO_ENABLED
 		.extra1		= SYSCTL_ZERO,
 		.extra2		= SYSCTL_INT_MAX,
+#endif
 	},
+#ifndef HAVE_REMOVE_SENTINEL_FROM_CTL_TABLE
+	{ }
+#endif
 };
 
 /*
@@ -483,6 +491,7 @@ static void iw_cm_check_wildcard(struct
 			cm4_outaddr->sin_addr = cm4_addr->sin_addr;
 		}
 	} else {
+#if IS_ENABLED(CONFIG_IPV6)
 		struct sockaddr_in6 *pm6_addr = (struct sockaddr_in6 *)pm_addr;
 
 		if (ipv6_addr_type(&pm6_addr->sin6_addr) == IPV6_ADDR_ANY) {
@@ -493,6 +502,7 @@ static void iw_cm_check_wildcard(struct
 
 			cm6_outaddr->sin6_addr = cm6_addr->sin6_addr;
 		}
+#endif
 	}
 }
 
