From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: net/sunrpc/xprtrdma/backchannel.c

---
 net/sunrpc/xprtrdma/backchannel.c | 28 ++++++++++++++++++++++++++--
 1 file changed, 26 insertions(+), 2 deletions(-)

diff --git a/net/sunrpc/xprtrdma/backchannel.c b/net/sunrpc/xprtrdma/backchannel.c
index xxxxxxx..xxxxxxx 100644
--- a/net/sunrpc/xprtrdma/backchannel.c
+++ b/net/sunrpc/xprtrdma/backchannel.c
@@ -11,10 +11,18 @@
 #include <linux/sunrpc/svc_rdma.h>
 
 #include "xprt_rdma.h"
+#ifdef HAVE_TRACE_RPCRDMA_H
 #include <trace/events/rpcrdma.h>
+#endif
 
 #undef RPCRDMA_BACKCHANNEL_DEBUG
 
+#if IS_ENABLED(CONFIG_SUNRPC_DEBUG)
+#ifndef RPCDBG_FACILITY
+#define RPCDBG_FACILITY    RPCDBG_TRANS
+#endif
+#endif
+
 /**
  * xprt_rdma_bc_setup - Pre-allocate resources for handling backchannel requests
  * @xprt: transport associated with these backchannel resources
@@ -27,10 +35,13 @@ int xprt_rdma_bc_setup(struct rpc_xprt *xprt, unsigned int reqs)
 	struct rpcrdma_xprt *r_xprt = rpcx_to_rdmax(xprt);
 
 	r_xprt->rx_buf.rb_bc_srv_max_requests = RPCRDMA_BACKWARD_WRS >> 1;
+#ifdef HAVE_TRACE_RPCRDMA_H
 	trace_xprtrdma_cb_setup(r_xprt, reqs);
+#endif
 	return 0;
 }
 
+
 /**
  * xprt_rdma_bc_maxpayload - Return maximum backchannel message size
  * @xprt: transport
@@ -78,7 +89,9 @@ static int rpcrdma_bc_marshal_reply(struct rpc_rqst *rqst)
 				      &rqst->rq_snd_buf, rpcrdma_noch_pullup))
 		return -EIO;
 
+#ifdef HAVE_TRACE_RPCRDMA_H
 	trace_xprtrdma_cb_reply(r_xprt, rqst);
+#endif
 	return 0;
 }
 
@@ -138,7 +151,6 @@ void xprt_rdma_bc_destroy(struct rpc_xprt *xprt, unsigned int reqs)
 		spin_unlock(&xprt->bc_pa_lock);
 
 		rpcrdma_req_destroy(rpcr_to_rdmar(rqst));
-
 		spin_lock(&xprt->bc_pa_lock);
 	}
 	spin_unlock(&xprt->bc_pa_lock);
@@ -239,7 +251,6 @@ void rpcrdma_bc_receive_call(struct rpcrdma_xprt *r_xprt,
 	rqst = rpcrdma_bc_rqst_get(r_xprt);
 	if (!rqst)
 		goto out_overflow;
-
 	rqst->rq_reply_bytes_recvd = 0;
 	rqst->rq_xid = *p;
 
@@ -258,14 +269,27 @@ void rpcrdma_bc_receive_call(struct rpcrdma_xprt *r_xprt,
 	 */
 	req = rpcr_to_rdmar(rqst);
 	req->rl_reply = rep;
+#ifdef HAVE_TRACE_RPCRDMA_H
 	trace_xprtrdma_cb_call(r_xprt, rqst);
+#endif
 
 	/* Queue rqst for ULP's callback service */
 	bc_serv = xprt->bc_serv;
 	xprt_get(xprt);
+#ifdef HAVE_SVC_SERV_SV_CB_LIST_LWQ
 	lwq_enqueue(&rqst->rq_bc_list, &bc_serv->sv_cb_list);
+#endif
+#ifdef HAVE_SVC_SERV_SV_CB_LIST_LIST_HEAD
+ 	spin_lock(&bc_serv->sv_cb_lock);
+ 	list_add(&rqst->rq_bc_list, &bc_serv->sv_cb_list);
+ 	spin_unlock(&bc_serv->sv_cb_lock);
+#endif
 
+#ifdef HAVE_SVC_POOL_WAKE_IDLE_THREAD
 	svc_pool_wake_idle_thread(&bc_serv->sv_pools[0]);
+#else
+ 	wake_up(&bc_serv->sv_cb_waitq);
+#endif
 
 	r_xprt->rx_stats.bcall_count++;
 	return;
