From: Tom Wu <tomwu@nvidia.com>
Subject: [PATCH] BACKPORT: include/trace/events/rpcrdma.h

Signed-off-by: Tom Wu <tomwu@nvidia.com>
---
 include/trace/events/rpcrdma.h | 2 ++
 1 file changed, 2 insertions(+)

--- a/include/trace/events/rpcrdma.h
+++ b/include/trace/events/rpcrdma.h
@@ -10,6 +10,8 @@
 #if !defined(_TRACE_RPCRDMA_H) || defined(TRACE_HEADER_MULTI_READ)
 #define _TRACE_RPCRDMA_H
 
+#include "../../../compat/config.h"
+
 #include <linux/scatterlist.h>
 #include <linux/sunrpc/rpc_rdma_cid.h>
 #include <linux/tracepoint.h>
@@ -1462,6 +1464,7 @@ TRACE_EVENT(svcrdma_encode_wseg,
 	)
 );
 
+#ifdef HAVE_SVC_RDMA_PCL
 TRACE_EVENT(svcrdma_decode_rseg,
 	TP_PROTO(
 		const struct rpc_rdma_cid *cid,
@@ -1534,6 +1537,7 @@ TRACE_EVENT(svcrdma_decode_wseg,
 		(unsigned long long)__entry->offset, __entry->handle
 	)
 );
+#endif
 
 DECLARE_EVENT_CLASS(svcrdma_error_event,
 	TP_PROTO(
