From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: net/sunrpc/xprtrdma/xprt_rdma.h

---
 net/sunrpc/xprtrdma/xprt_rdma.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/net/sunrpc/xprtrdma/xprt_rdma.h b/net/sunrpc/xprtrdma/xprt_rdma.h
index xxxxxxx..xxxxxxx 100644
--- a/net/sunrpc/xprtrdma/xprt_rdma.h
+++ b/net/sunrpc/xprtrdma/xprt_rdma.h
@@ -56,7 +56,9 @@
 #include <linux/sunrpc/rpc_rdma_cid.h> 	/* completion IDs */
 #include <linux/sunrpc/rpc_rdma.h> 	/* RPC/RDMA protocol */
 #include <linux/sunrpc/xprtrdma.h> 	/* xprt parameters */
+#ifdef HAVE_RPCRDMA_RN_REGISTER
 #include <linux/sunrpc/rdma_rn.h>	/* removal notifications */
+#endif
 
 #define RDMA_RESOLVE_TIMEOUT	(5000)	/* 5 seconds */
 #define RDMA_CONNECT_RETRY_MAX	(2)	/* retries if no listener backlog */
@@ -93,7 +95,9 @@ struct rpcrdma_ep {
 	struct rpcrdma_connect_private
 				re_cm_private;
 	struct rdma_conn_param	re_remote_cma;
+#ifdef HAVE_RPCRDMA_RN_REGISTER
 	struct rpcrdma_notification	re_rn;
+#endif
 	int			re_receive_count;
 	unsigned int		re_max_requests; /* depends on device */
 	unsigned int		re_inline_send;	/* negotiated */
@@ -431,7 +435,9 @@ struct rpcrdma_xprt {
 	struct rpcrdma_ep	*rx_ep;
 	struct rpcrdma_buffer	rx_buf;
 	struct delayed_work	rx_connect_worker;
+#ifdef HAVE_XPRT_RECONNECT_DELAY
 	struct rpc_timeout	rx_timeout;
+#endif
 	struct rpcrdma_stats	rx_stats;
 };
 
@@ -459,6 +465,7 @@ extern int xprt_rdma_pad_optimize;
  */
 extern unsigned int xprt_rdma_memreg_strategy;
 
+
 /*
  * Endpoint calls - xprtrdma/verbs.c
  */
