From: Valentine Fatiev <valentinef@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/nvme/host/apple.c

---
 drivers/nvme/host/apple.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

--- a/drivers/nvme/host/apple.c
+++ b/drivers/nvme/host/apple.c
@@ -874,7 +874,12 @@ static void apple_nvme_disable(struct ap
 	}
 }
 
+#ifdef HAVE_BLK_MQ_OPS_TIMEOUT_1_PARAM
 static enum blk_eh_timer_return apple_nvme_timeout(struct request *req)
+#else
+static enum blk_eh_timer_return apple_nvme_timeout(struct request *req,
+						   bool reserved)
+#endif
 {
 	struct apple_nvme_iod *iod = blk_mq_rq_to_pdu(req);
 	struct apple_nvme_queue *q = iod->q;
@@ -1251,7 +1256,9 @@ static int apple_nvme_alloc_tagsets(stru
 	anv->admin_tagset.timeout = NVME_ADMIN_TIMEOUT;
 	anv->admin_tagset.numa_node = NUMA_NO_NODE;
 	anv->admin_tagset.cmd_size = sizeof(struct apple_nvme_iod);
+#ifdef HAVE_BLK_MQ_F_NO_SCHED
 	anv->admin_tagset.flags = BLK_MQ_F_NO_SCHED;
+#endif
 	anv->admin_tagset.driver_data = &anv->adminq;
 
 	ret = blk_mq_alloc_tag_set(&anv->admin_tagset);
@@ -1275,7 +1282,9 @@ static int apple_nvme_alloc_tagsets(stru
 	anv->tagset.timeout = NVME_IO_TIMEOUT;
 	anv->tagset.numa_node = NUMA_NO_NODE;
 	anv->tagset.cmd_size = sizeof(struct apple_nvme_iod);
+#ifdef HAVE_BLK_MQ_F_SHOULD_MERGE
 	anv->tagset.flags = BLK_MQ_F_SHOULD_MERGE;
+#endif
 	anv->tagset.driver_data = &anv->ioq;
 
 	ret = blk_mq_alloc_tag_set(&anv->tagset);
@@ -1516,7 +1525,11 @@ static int apple_nvme_probe(struct platf
 		goto put_dev;
 	}
 
+#ifdef HAVE_BLK_MQ_ALLOC_QUEUE
 	anv->ctrl.admin_q = blk_mq_alloc_queue(&anv->admin_tagset, NULL, NULL);
+#else
+	anv->ctrl.admin_q = blk_mq_init_queue(&anv->admin_tagset);
+#endif
 	if (IS_ERR(anv->ctrl.admin_q)) {
 		ret = -ENOMEM;
 		goto put_dev;
