From: Maor Gottlieb <maorg@nvidia.com>
Subject: [PATCH] BACKPORT: drivers/infiniband/core/ucaps.c

---
 drivers/infiniband/core/ucaps.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- a/drivers/infiniband/core/ucaps.c
+++ b/drivers/infiniband/core/ucaps.c
@@ -29,7 +29,11 @@ static const char *ucap_names[RDMA_UCAP_
 	[RDMA_UCAP_MLX5_CTRL_OTHER_VHCA] = "mlx5_perm_ctrl_other_vhca"
 };
 
+#ifdef HAVE_DEVNODE_GET_CONST_DEVICE
 static char *ucaps_devnode(const struct device *dev, umode_t *mode)
+#else
+static char *ucaps_devnode(struct device *dev, umode_t *mode)
+#endif
 {
 	if (mode)
 		*mode = 0600;
@@ -37,7 +41,7 @@ static char *ucaps_devnode(const struct
 	return kasprintf(GFP_KERNEL, "infiniband/%s", dev_name(dev));
 }
 
-static const struct class ucaps_class = {
+static struct class ucaps_class = {
 	.name = "infiniband_ucaps",
 	.devnode = ucaps_devnode,
 };
