From: Lama Kayal <lkayal@nvidia.com>
Subject: [PATCH] BACKPORT: net/mlxdevm/mlxdevm_ext.h

---
 net/mlxdevm/mlxdevm_ext.h | 4 ++++
 1 file changed, 4 insertions(+)

--- a/net/mlxdevm/mlxdevm_ext.h
+++ b/net/mlxdevm/mlxdevm_ext.h
@@ -1,6 +1,9 @@
 #ifndef _MLXDEVM_EXT_H
 #define _MLXDEVM_EXT_H
 
+#include "netlink_gen.h"
+#include "../../include/net/mlxdevm.h"
+#ifdef HAVE_DEVLINK_HAS_RATE_FUNCTIONS
 struct mlxdevm_rate_node {
 	struct list_head list;
 	char *name;
@@ -10,3 +13,4 @@ struct mlxdevm_rate_node {
 	u32 tc_bw[IEEE_8021QAZ_MAX_TCS];
 };
 #endif
+#endif
